### PR TITLE
CBG-1966 enable staticcheck

### PR DESCRIPTION
**Pull request by torcolvin**
- remove kvPoolSize empty check (setup in GetGoCBConnString
- replace math/rand with crypto/rand since math/rand.Seed is deprecated since 1.20 for crypto/rand.Seed
- remove random seed on init since go 1.20, which is now builtin, rather than Seed(1)
- remove potential nil checks after we acquire locks in ActiveReplicator
- remove explicit set to benchmark
- check error before defer foo.Close()

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/2141/

### CODE REVIEW

File: auth/oidc_test.go [hunk 1]
```diff
@@ -1255,7 +1255,6 @@ func TestJWTRolesChannels(t *testing.T) {
 				require.Equal(t, base.SetFromArray(login.expectedChannels), user.Channels().AsSet())
 
 				require.Greater(t, user.JWTLastUpdated(), lastUpdateTime)
-				lastUpdateTime = user.JWTLastUpdated()
```
Comment (bbrks) [inline auth/oidc_test.go:1258]: This doesn't quite look right - comparing greater than `0` seems unintentional.

File: auth/password_hash_test.go [hunk 2]
```diff
@@ -35,7 +35,6 @@ func BenchmarkBcryptCostTimes(b *testing.B) {
 
 	for i := minCostToTest; i < maxCostToTest; i++ {
 		b.Run(fmt.Sprintf("cost%d", i), func(bn *testing.B) {
```
Comment (bbrks) [inline auth/password_hash_test.go:37]: Benchmark looks wrong as-is anyway (no b.N loop)

Maybe we should just swap this to a test that logs the values, as we get the same one-off benefit even without assertions.

```go
// TestBcryptCostTimes will output the time it takes to hash a password with each bcrypt cost value
func TestBcryptCostTimes(t *testing.T) {
	// Little value in running this regularly. Might be useful for one-off informational purposes
	t.Skip("Test disabled")

	minCostToTest := bcrypt.DefaultCost
	maxCostToTest := bcrypt.DefaultCost + 5

	for i := minCostToTest; i < maxCostToTest; i++ {
		t.Run(fmt.Sprintf("cost%d", i), func(t *testing.T) {
			startTime := time.Now()
			_, err := bcrypt.GenerateFromPassword([]byte("hunter2"), i)
			duration := time.Since(startTime)

			t.Logf("bcrypt.GenerateFromPassword with cost %d took: %v", i, duration)
			assert.NoError(t, err)
		})
	}
}
```

```
=== RUN   TestBcryptCostTimes
--- PASS: TestBcryptCostTimes (2.11s)
=== RUN   TestBcryptCostTimes/cost10
    password_hash_test.go:42: bcrypt.GenerateFromPassword with cost 10 took: 68.808833ms
    --- PASS: TestBcryptCostTimes/cost10 (0.07s)
=== RUN   TestBcryptCostTimes/cost11
    password_hash_test.go:42: bcrypt.GenerateFromPassword with cost 11 took: 136.465541ms
    --- PASS: TestBcryptCostTimes/cost11 (0.14s)
=== RUN   TestBcryptCostTimes/cost12
    password_hash_test.go:42: bcrypt.GenerateFromPassword with cost 12 took: 271.3435ms
    --- PASS: TestBcryptCostTimes/cost12 (0.27s)
=== RUN   TestBcryptCostTimes/cost13
    password_hash_test.go:42: bcrypt.GenerateFromPassword with cost 13 took: 546.19325ms
    --- PASS: TestBcryptCostTimes/cost13 (0.55s)
=== RUN   TestBcryptCostTimes/cost14
    password_hash_test.go:42: bcrypt.GenerateFromPassword with cost 14 took: 1.089720583s
    --- PASS: TestBcryptCostTimes/cost14 (1.09s)
PASS
```

File: base/collection.go [hunk 3]
```diff
@@ -63,10 +63,6 @@ func GetGoCBv2Bucket(ctx context.Context, spec BucketSpec) (*GocbV2Bucket, error
 		RetryStrategy:  gocb.NewBestEffortRetryStrategy(nil),
 	}
 
-	if spec.KvPoolSize > 0 {
-		// TODO: Equivalent of kvPoolSize in gocb v2?
-	}
```
Comment (bbrks) [inline base/collection.go:68]: ~Are we actually OK removing this? Or should we file a ticket if we think it's not implemented?~

I read the PR description highlighting `GetGoCBConnString`. All good.

File: db/active_replicator_pull.go [hunk 4]
```diff
@@ -40,10 +40,6 @@ func (apr *ActivePullReplicator) Start(ctx context.Context) error {
 	apr.lock.Lock()
 	defer apr.lock.Unlock()
 
-	if apr == nil {
-		return fmt.Errorf("nil ActivePullReplicator, can't start")
-	}
```
Comment (bbrks) [inline db/active_replicator_pull.go:45]: Wouldn't a more appropriate fix be to move this to above the lock, or is it guaranteed that `apr` can't be nil?

File: db/active_replicator_pull.go [hunk 5]
```diff
@@ -158,10 +154,6 @@ func (apr *ActivePullReplicator) _subChanges(collectionIdx *int, since string) e
 func (apr *ActivePullReplicator) Complete() {
 	base.TracefCtx(apr.ctx, base.KeyReplicate, "ActivePullReplicator.Complete()")
 	apr.lock.Lock()
-	if apr == nil {
-		apr.lock.Unlock()
-		return
-	}
```
Comment (bbrks) [inline db/active_replicator_pull.go:164]: Similarly, I'm not sure of the original motivation for this guard? Maybe being overly defensive?

File: db/attachment_test.go [hunk 6]
```diff
@@ -10,12 +10,12 @@ package db
 
 import (
 	"context"
+	"crypto/rand"
```
Comment (bbrks) [inline db/attachment_test.go]: This file/`TestLargeAttachments` doesn't seem to _need_ cryptographic randomness? What's the motivation for the change here?

File: db/crud.go [hunk 7]
```diff
@@ -413,10 +413,10 @@ func (db *DatabaseCollectionWithUser) GetDelta(ctx context.Context, docID, fromR
...
 			db.dbStats().DeltaSync().DeltaCacheHit.Add(1)
 			return fromRevision.Delta, nil, nil
-		} else {
-			// TODO: Recurse and merge deltas when gen(revCacheDelta.toRevID) < gen(toRevId)
-			// until then, fall through to generating delta for given rev pair
-		}
+		} // else {
+		// TODO: Recurse and merge deltas when gen(revCacheDelta.toRevID) < gen(toRevId)
+		// until then, fall through to generating delta for given rev pair
+		//	}
```
Comment (bbrks) [inline db/crud.go]: I'd be OK removing this completely. Storage/memory vs. CPU tradeoff.

File: rest/multipart_test.go [hunk 8]
```diff
@@ -163,7 +164,8 @@ func TestWriteJSONPart(t *testing.T) {
 	// a body larger than 300 bytes to test writeJSONPart with compression=true and compression=false
 	mockFakeBody := func() db.Body {
 		bytes := make([]byte, 139)
-		rand.Read(bytes)
+		_, err := rand.Read(bytes)
+		require.NoError(t, err)
```
Comment (bbrks) [inline rest/multipart_test.go]: Doesn't need crypto rand

**Review by bbrks — CHANGES_REQUESTED** [review]
This one has some questionable changes/lint warnings. See inline for details.

File: db/active_replicator_pull.go (hunk 4 above)
Comment (torcolvin) [reply db/active_replicator_pull.go:45]: I think because the code works as is, that this was just overly defensive programming and broken in a refactor.

File: db/active_replicator_pull.go (hunk 5 above)
Comment (torcolvin) [reply db/active_replicator_pull.go:164]: I think this was overly defensive programming, and is not hit currently.

File: db/attachment_test.go (hunk 6 above)
Comment (torcolvin) [reply db/attachment_test.go]: https://github.com/golang/go/issues/20661 officially deprecates `math/rand` for `crypto/rand`. The answer seems to be that people are worried that you'll use the wrong one by mistake.

I don't know that we need cryptographic randomness anywhere, but I could see making this mistake. Most tools (goimports, etc) will add cryptographic rand anyway for new code.

We could make an exception for this because I don't think that we currently need cryptographic randomness anywhere, but I don't think this is necessary.

File: auth/oidc_test.go (hunk 1 above)
Comment (torcolvin) [reply auth/oidc_test.go:1258]: I think this is right, because it is checking that this value got updated - are you suggesting I remove this check, or move `lastUpdated` time lower than `user.SetJWTLastUpdated(time.Now())` and use `AssertTimestampGreaterThan`?

Comment (torcolvin) [reply auth/oidc_test.go:1258]: Nevermind, this doesn't get updated.

File: db/attachment_test.go (hunk 6 above)
Comment (bbrks) [reply db/attachment_test.go]: Looks like the intention is to switch to a locally scoped `math.Rand.Read()` if you don't need cryptographically secure randomness. We should do that here

**Review by bbrks — APPROVED** [review]
